go/pserver/service.go

```diff
@@ -142,8 +177,51 @@ func (s *Service) GetParam(name string, parameter *Parameter) error {
 
 // Save tells the parameter server to save parameters.
 func (s *Service) Save(path string, dummy *int) error {
+	//FIXME: checkpoint is only used by pserver
+	// and has a constant path of */checkpoints/{pserver_idx}*
 	<-s.initialized
-
-	// TODO
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	var paramWithConfig ParameterWithConfig
+	for name, opt := range s.optMap {
+		paramWithConfig.Param.Name = name
+		paramWithConfig.Param.ElementType = opt.elementType
+		paramWithConfig.Param.Content = opt.GetWeights()
+		paramWithConfig.State = opt.GetStates()
+		content, err := GetBytes(paramWithConfig)
+		if err != nil {
+			log.Errorln(err)
+		}
+		ck := checkpoint{}
+		h := md5.New()
+		ck.Md5sum = hex.EncodeToString(h.Sum(content))
+		ck.Timestamp = time.Now().String()
+		ck.Uuid = checkpoint_path + strconv.Itoa(s.idx)
+		ckbytes, err := GetBytes(ck)
+		if err != nil {
+			log.Errorln(err)
+		}
+		// TODO: according design doc, need to save Uuid to etcd in json format
+		// {\"Uuid\": [UUID], \"md5\", \"MD5 sum\", \"Timestamp\": xxxx}
+		log.Infof("parameter checkpoint %s", ckbytes)
+
+		if _, err = os.Stat(ck.Uuid); os.IsNotExist(err) {
+			log.Info("checkpoint not exists.")
+		} else {
+			err = os.Remove(ck.Uuid)
+			log.Infof("remove %s", ck.Uuid)
+		}
+		f, err := os.Create(ck.Uuid)
```

Comment:
It will create so many files for each paramter. Following the design doc, we will only have one checkpoint file named UUID?